uniffi_bindgen/src/bindings/kotlin/templates/ObjectTemplate.kt

```diff
@@ -58,8 +58,16 @@ class {{ type_name }}(
     override suspend fun {{ meth.name()|fn_name }}({%- call kt::arg_list_protocol(meth) -%}){% match meth.return_type() %}{% when Some with (return_type) %}: {{ return_type|type_name }}{% when None %}{% endmatch %} {
         class Waker: RustFutureWaker {
             override fun callback(envCStructure: RustFutureWakerEnvironmentCStructure?) {
-                val hash = envCStructure!!.hash
-                val env = _UniFFILib.FUTURE_WAKER_ENVIRONMENTS.get(hash)!!
+                if (envCStructure == null) {
+                    return;
```

Comment:
I should have commented a couple of lines lower. My bad

Reply:
OK, so a consecutive call to `wake` should wait on the former to finish. Put it diferently, the $(n+1)^\mathrm{th}$ call should wait on the $n^\mathrm{th}$ call to finish. We can keep the logic that is already present to return when the env has been removed, but we should wait to see if the env will be “re-inserted” or not: If it's re-inserted, it means the future can be polled again because it has returned `Pending`, otherwise —if it has not been re-inserted—, it means the future is resolved (it has returned `Ready`), and we can `return` properly.

[Polling a `Ready` `Future` is an UB](https://doc.rust-lang.org/std/future/trait.Future.html#panics), so returning from the waker gracefully seems appropriate here.

Reply:
OK this is solved in Kotlin. I'm using a `Semaphore` to allow only one polling at a time in the waker. The presence of the waker's environment tells about the state of the future. It works like a charm.

Now let's do it for Swift and Python :-).